### PR TITLE
hot reload for views

### PR DESCRIPTION
**Pull request by munerf**

### CODE REVIEW

File: ReactViewControl/ViewModuleContainer.cs [hunk 1]
```diff
@@ -80,7 +96,14 @@ public abstract class ViewModuleContainer : IViewModule {
             var entriesFilePath = VirtualPathUtility.GetDirectory(MainJsSource) + Path.GetFileNameWithoutExtension(MainJsSource) + extension;
             var resource = entriesFilePath.Split(new[] { ResourceUrl.PathSeparator }, StringSplitOptions.RemoveEmptyEntries);
 
-            var stream = ResourcesManager.TryGetResourceWithFullPath(resource.First(), resource);
+            Stream stream;
+            if (liveReload) {
+                var basePath = Path.GetDirectoryName(Source);
+                stream = File.OpenRead($"{basePath}\\{resource.Last()}");
```
Comment (alvesmiguel1) [inline ReactViewControl/ViewModuleContainer.cs:102]: use Path.Combine instead

Comment (alvesmiguel1) [reply ReactViewControl/ViewModuleContainer.cs:102]: We should have something like:
```
using (var stream = liveReload ? File.OpenRead(...) : ResourcesManager.Try(..)) {
(....)
}
```

**Comment by alvesmiguel1** [comment]
Adding @Jmales to have an extra pair of eyes

File: ReactViewControl/IViewModule.cs [hunk 2]
```diff
@@ -6,9 +6,9 @@ public interface IViewModule {
 
         string MainJsSource { get; }
 
-        string[] DependencyJsSources { get; }
+        string[] DependencyJsSources(bool useCache);
```
Comment (alvesmiguel1) [inline ReactViewControl/IViewModule.cs:9]: You don't need to use this approach. You can modify ViewModuleContainer's Bind method to receive a parameter corresponding to the hot reload enabled. You should also set a Function in ViewModuleContainer that is invoked when we get DependencyJsSources (or CssSources). E.g.

```
private Function<string[]> func = DependencyJsSourcesCache.Value;

string[] IViewModule.DependencyJsSources => func.Invoke();

IViewModule.Bind(...) {
  if (hotReload) {
    func = () => GetDependenciesFromEntriesFile(... , hotReload: true);
  }
}
```

Comment (alvesmiguel1) [reply ReactViewControl/IViewModule.cs:9]: This works for plugins as well

File: ReactViewControl/ViewModuleContainer.cs [hunk 3]
```diff
@@ -76,18 +78,24 @@ public abstract class ViewModuleContainer : IViewModule {
...
-            var stream = ResourcesManager.TryGetResourceWithFullPath(resource.First(), resource);
-            if (stream != null) {
-                using (var reader = new StreamReader(stream)) {
-                    var allEntries = reader.ReadToEnd();
-                    if (allEntries != null && allEntries != string.Empty) {
-                        return allEntries.Split(new[] { "\n" }, StringSplitOptions.RemoveEmptyEntries);
+            var basePath = Path.GetDirectoryName(Source);
+
+            using (var stream = hotReloadEnabled
+                ? File.OpenRead(Path.Combine(basePath, resource.Last()))
```
[the remark this answers is not in the document]
Comment (alvesmiguel1) [reply ReactViewControl/ViewModuleContainer.cs:87]: This works properly for large sources like we have it in SS?

File: ReactViewControl/ViewModuleContainer.cs [hunk 4]
```diff
@@ -15,16 +15,23 @@ public abstract class ViewModuleContainer : IViewModule {
...
         private IChildViewHost childViewHost;
 
+
         public ViewModuleContainer() {
             DependencyJsSourcesCache = new Lazy<string[]>(() => GetDependenciesFromEntriesFile(JsEntryFileExtension));
             CssSourcesCache = new Lazy<string[]>(() => GetDependenciesFromEntriesFile(CssEntryFileExtension));
 
             frame = new FrameInfo("dummy");
+            dependencyJsSourcesDelegate = () => DependencyJsSourcesCache.Value;
+            dependencyJsSourcesDelegate = () => CssSourcesCache.Value;
```
Comment (alvesmiguel1) [inline ReactViewControl/ViewModuleContainer.cs:25]: dependencyCssSourcesDelegate

File: ReactViewControl/ViewModuleContainer.cs [hunk 5]
```diff
@@ -15,16 +15,23 @@ public abstract class ViewModuleContainer : IViewModule {
...
         }
 
-        private Lazy<string[]> DependencyJsSourcesCache { get; }
+        public Lazy<string[]> DependencyJsSourcesCache { get; }
```
Comment (alvesmiguel1) [inline ReactViewControl/ViewModuleContainer.cs:28]: public??

File: ReactViewControl/ViewModuleContainer.cs [hunk 6]
```diff
@@ -52,17 +59,23 @@ public abstract class ViewModuleContainer : IViewModule {
 
         string[] IViewModule.Events => Events;
 
-        string[] IViewModule.DependencyJsSources => DependencyJsSourcesCache.Value;
+        string[] IViewModule.DependencyJsSources => dependencyJsSourcesDelegate.Invoke();
```
Comment (alvesmiguel1) [inline ReactViewControl/ViewModuleContainer.cs:62]: Please use `dependencyJsSourcesDelegate?.Invoke() ?? DependencyJsSourcesCache.Value` and remove the constructor assignment of `dependencyJsSourcesDelegate`. Same for CSS

File: Example/tsconfig.json [hunk 7]
```diff
@@ -16,6 +16,6 @@
     "compileOnSave": false,
     "buildOnSave": false,
     "exclude": [
-        "dist"
+        "dist", "Generated"
```
[the remark this answers is not in the document]
Comment (alvesmiguel1) [reply Example/tsconfig.json:19]: Can you remove it if not necessary?

File: Example/Example.csproj [hunk 8]
```diff
@@ -1,7 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="..\packages\ViewGenerator.1.0.293\build\ViewGenerator.props" Condition="Exists('..\packages\ViewGenerator.1.0.293\build\ViewGenerator.props')" />
-  <Import Project="..\packages\ViewPacker.1.0.12\build\ViewPacker.props" Condition="Exists('..\packages\ViewPacker.1.0.12\build\ViewPacker.props')" />
+  <Import Project="..\packages\ViewGenerator.1.0.296\build\ViewGenerator.props" Condition="Exists('..\packages\ViewGenerator.1.0.296\build\ViewGenerator.props')" />
```
Comment (alvesmiguel1) [inline Example/Example.csproj:3]: Actually, you should revert these upgrades (of ViewGenerator and ViewPacker) in Example. The azure deployment needs to generate the packages first, only then we can upgrade

Comment (alvesmiguel1) [reply Example/Example.csproj:3]: In packages.config as well

File: ViewGenerator/ViewGenerator.nuspec [hunk 9]
```diff
@@ -2,7 +2,7 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd">
```
[the remark this answers is not in the document]
Comment (alvesmiguel1) [reply ViewGenerator/ViewGenerator.nuspec:2]: You missed this one

File: ReactViewControl/ViewModuleContainer.cs [hunk 10]
```diff
@@ -52,17 +56,23 @@ public abstract class ViewModuleContainer : IViewModule {
...
         KeyValuePair<string, object>[] IViewModule.PropertiesValues => PropertiesValues;
 
-        void IViewModule.Bind(IFrame frame, IChildViewHost childViewHost) {
+        void IViewModule.Bind(IFrame frame, bool enableHotReload, IChildViewHost childViewHost) {
             frame.CustomResourceRequestedHandler += this.frame.CustomResourceRequestedHandler;
             frame.ExecutionEngine.MergeWorkload(this.frame.ExecutionEngine);
             this.frame = frame;
             this.childViewHost = childViewHost;
+
+            if (enableHotReload) {
```
Comment (joaompneves) [inline ReactViewControl/ViewModuleContainer.cs:71]: just store the value of the enableHotReload ... as a class field and use that on the GetDependenciesFromEntriesFile no more changes needed

Comment (joaompneves) [reply ReactViewControl/ViewModuleContainer.cs:71]: btw enableHotReload -> isHotReloadEnabled

File: ReactViewControl/ViewModuleContainer.cs [hunk 11]
```diff
@@ -52,17 +56,23 @@ public abstract class ViewModuleContainer : IViewModule {
...
-        string[] IViewModule.DependencyJsSources => DependencyJsSourcesCache.Value;
+        string[] IViewModule.DependencyJsSources => getDependencyJsSources?.Invoke() ?? DependencyJsSourcesCache.Value;
 
-        string[] IViewModule.CssSources => CssSourcesCache.Value;
+        string[] IViewModule.CssSources => getCssSources?.Invoke() ?? CssSourcesCache.Value;
 
         KeyValuePair<string, object>[] IViewModule.PropertiesValues => PropertiesValues;
 
-        void IViewModule.Bind(IFrame frame, IChildViewHost childViewHost) {
+        void IViewModule.Bind(IFrame frame, bool enableHotReload, IChildViewHost childViewHost) {
```
Comment (joaompneves) [inline ReactViewControl/ViewModuleContainer.cs:65]: isHotReloadEnabled

File: ReactViewControl/ViewModuleContainer.cs [hunk 12]
```diff
@@ -76,18 +86,23 @@ public abstract class ViewModuleContainer : IViewModule {
...
             var entriesFilePath = VirtualPathUtility.GetDirectory(MainJsSource) + Path.GetFileNameWithoutExtension(MainJsSource) + extension;
             var resource = entriesFilePath.Split(new[] { ResourceUrl.PathSeparator }, StringSplitOptions.RemoveEmptyEntries);
 
-            var stream = ResourcesManager.TryGetResourceWithFullPath(resource.First(), resource);
-            if (stream != null) {
-                using (var reader = new StreamReader(stream)) {
-                    var allEntries = reader.ReadToEnd();
-                    if (allEntries != null && allEntries != string.Empty) {
-                        return allEntries.Split(new[] { "\n" }, StringSplitOptions.RemoveEmptyEntries);
+            using (var stream = hotReloadEnabled
```
Comment (joaompneves) [inline ReactViewControl/ViewModuleContainer.cs:93]: hotReloadEnabled
            if (stream != null) {	                ? File.OpenRead(Path.Combine(Path.GetDirectoryName(Source), resource.Last()))
                using (var reader = new StreamReader(stream)) {	                : ResourcesManager.TryGetResourceWithFullPath(resource.First(), resource)


move into a method GetResourceStream

File: ReactViewControl/ViewModuleContainer.cs [hunk 13]
```diff
@@ -23,8 +23,12 @@ public abstract class ViewModuleContainer : IViewModule {
         }
 
         private Lazy<string[]> DependencyJsSourcesCache { get; }
+
         private Lazy<string[]> CssSourcesCache { get; }
 
+        private Func<string[]> getDependencyJsSources;
```
Comment (joaompneves) [inline ReactViewControl/ViewModuleContainer.cs:29]: delete these

File: ReactViewControl/ReactViewRender.cs [hunk 14]
```diff
@@ -125,6 +125,10 @@ internal partial class ReactViewRender : UserControl, IDisposable, IChildViewHos
             }
         }
 
+        public bool EnableHotReload { get; set; }
```
Comment (joaompneves) [inline ReactViewControl/ReactViewRender.cs:128]: remove ... just check the url

**Pull request merged**